File: testing_denoise/NLmeans.m
delete(gcp)
parpool(4);

dirname = './testing_image/';
savedirname = './result/find_interval/';
MyFolderInfo = dir(dirname);

parfor j=3:length(MyFolderInfo)
    %var=[1,3,5,7,9,11,13,17,33,65,129]
    %var=[1,3,5,7,9,11,13,17,33,65,129]
    var=[2^1,2^2,2^3,2^4,2^5,2^6,2^7,2^8,2^9,2^10,2^11]

    showimg=zeros(256,256,12);
    img = imread(strcat(dirname, MyFolderInfo(j).name));
    img = im2double(img);
    showimg(:,:,1)=img;
    
    for i = 1:11
        %dst = NLmeansfilter(img,var(i),1,1);
        %dst = NLmeansfilter(img,1,var(i),1);
        dst = NLmeansfilter(img,1,1,var(i));

        showimg(:,:,i+1)=dst;
        j
        i
    end
    
    figure;
    title('h');
    for i = 1:12
        subplot(3,4,i), imshow(showimg(:,:,i));
        
        if i==1
           title('original');
        else
           title(num2str(var(i-1)));
        end
    end
    tightfig;
    saveas(gcf,strcat(savedirname,MyFolderInfo(j).name,'_NLMeans_h.png'));
end

File: testing_denoise/NL-means.py
import numpy as np
import cv2
from matplotlib import pyplot as plt
import os

dir_name = "./testing_image/"
images = os.listdir(dir_name)

title = "patchsize"
var = [1,2,4,8,16,32,64,128,256,512,1024]
#var = [2,4,8,16,32,64,128,256,512,1024,2048] #h
for i, imname in enumerate(images):
  f, a = plt.subplots(3, 4, figsize=(7,7))
  f.suptitle(title)
  showimg = []
  img = cv2.imread(dir_name+imname)
  showimg.append(img)

  for j in var:
    dst = cv2.fastNlMeansDenoising(img,None,10,j,7)
    showimg.append(dst)

  for j in range(len(var)+1):
    if j!=0:
      a[j//4][j%4].set_title(var[j-1])
    a[j//4][j%4].imshow(showimg[j], cmap="gray")
    a[j//4][j%4].set_xticks(()); a[j//4][j%4].set_yticks(());
      
  plt.show()
  #f.savefig("./result/find_interval/"+imname+"NLmeans_"+title+".png")
